lib/ruleHelper.js

```diff
@@ -76,15 +77,77 @@ RuleHelper.prototype = {
         } else if (expression.type === "CallExpression") {
             allowed = this.isAllowedCallExpression(expression.callee, escapeObject.methods || VALID_UNWRAPPERS);
         } else if (expression.type === "BinaryExpression") {
-            allowed = ((this.allowedExpression(expression.left, escapeObject))
-            && (this.allowedExpression(expression.right, escapeObject)));
+            allowed = ((this.allowedExpression(expression.left, escapeObject, details))
+            && (this.allowedExpression(expression.right, escapeObject, details)));
         } else if (expression.type === "TSAsExpression") {
             // TSAsExpressions contain the raw javascript value in 'expression'
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (expression.type === "TypeCastExpression") {
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (Array.isArray(expression)) {
-            allowed = expression.every((e) => this.allowedExpression(e, escapeObject));
+            allowed = expression.every((e) => this.allowedExpression(e, escapeObject, details));
+        } else if (expression.type === "Identifier") {
+            // find declared variables and see which are literals
+            const scope = this.context.getScope(expression);
+            const variableInfo = scope.set.get(expression.name);
+
+            // let's see if this variable is allowed, by tracing it back through the code
+            if (variableInfo) {
```

Comment:
If we proceed with refactoring out this block into its own separate method, we could also reduce the nesting level of the resulting function by returning earlier with false if `variableInfo` is not defined or if `variableInfo.defs` is an empty array (which is definitely unexpected, not sure if it is technically possible to be honest).

It would be nice to also report that case with a more detailed message, but I'm afraid that if we do something like this:

```js
    isAllowedIdentifier(expression, escapeObject, details) {
      // find declared variables and see which are literals
      const scope = this.context.getScope(expression);
      const variableInfo = scope.set.get(expression.name);

      if (!variableInfo || variableInfo.defs.length === 0) {
        // details.message = `Variable ${expression.name} considered unsafe: variable initialization not found`;
        return false;
      }
      ...
```

This error message may change the behavior in a bunch of our tests (which would need to be changed) and I'm not sure that it would be really helpful from a user perspective, and so I think we can defer looking into improving the linting message for the "unable to track down the variable" case for now and just return earlier to make the `isAllowedIdentifier` as linear as possible.

Reply:
I took almost all of your feedback except:
- I'll make TODO note about the linter message from your last comment and will do this in a follow-up
- I don't want to rewrite too much logic in one go, so the switch/case rewrite should happen after this feature-work (imho). I filed #172.

Reply:
#176 is the other follow-up. Resolving this.